channeldb/channel.go

```diff
@@ -3216,6 +3216,280 @@ const (
+				tlv.MakePrimitiveRecord(
+					ActionKeyType, &action,
+				),
+			)
+			if err != nil {
+				return err
+			}
+
+			var actionBytes bytes.Buffer
+
+			err = actionStream.Encode(&actionBytes)
+			if err != nil {
+				return err
+			}
+			if err != nil {
+				return err
+			}
+
+			// We encode the record with a varint length followed by
+			// the _raw_ TLV bytes.
+			tlvLen := uint64(len(actionBytes.Bytes()))
+			if err := tlv.WriteVarInt(w, tlvLen, buf); err != nil {
+				return err
+			}
+
+			_, err = w.Write(actionBytes.Bytes())
+			if err != nil {
+				return err
+			}
+			err = SerializeHtlcs(w, htlcs...)
+			if err != nil {
+				return err
+			}
+
+		}
+
+		return nil
+	}
+
+	return tlv.NewTypeForEncodingErr(val, "map[string][]HTLC")
+}
+
+// HtlcActionDecoder is a custom TLV decoder for the htlcAction record.
+func HtlcActionDecoder(r io.Reader, val interface{}, buf *[8]byte,
+	l uint64) error {
+
+	if v, ok := val.(*map[string][]HTLC); ok {
+		// First, we'll decode the varint that encodes how many set IDs
+		// are encoded within the greater map.
+		numRecords, err := tlv.ReadVarInt(r, buf)
+		if err != nil {
+			return err
+		}
+		// Now that we know how many records we'll need to read, we can
+		// iterate and read them all out in series.
+		for i := uint64(0); i < numRecords; i++ {
+			// Read out the varint that encodes the size of this
+			// inner TLV record.
+			actionRecordSize, err := tlv.ReadVarInt(r, buf)
+			if err != nil {
+				return err
+			}
+
+			// Using this information, we'll create a new limited
+			// reader that'll return an EOF once the end has been
+			// reached so the stream stops consuming bytes.
+			actionTlvReader := io.LimitedReader{
+				R: r,
+				N: int64(actionRecordSize),
+			}
+
+			var action []byte
+
+			actionStream, err := tlv.NewStream(
+				tlv.MakePrimitiveRecord(
+					ActionKeyType, &action,
+				),
+			)
+
+			if err != nil {
+				return err
+			}
+
+			err = actionStream.Decode(&actionTlvReader)
+			if err != nil {
+				return err
+			}
+
+			htlcs, err := DeserializeHtlcs(r)
+			if err != nil {
+				return err
+			}
+			(*v)[string(action)] = htlcs
+		}
+
+		return nil
+	}
+
+	return tlv.NewTypeForDecodingErr(
+		val, "map[string][]HTLC", l, l,
+	)
+}
+
+// SerializeLocalCloseInfo writes out the passed set of LocalForceCloseInfo into
+// the passed writer.
+func SerializeLocalCloseInfo(w io.Writer, lc *LocalForceCloseInfo) error {
+	linkFailureErrorByte := []byte(lc.LinkFailureError)
+
+	tlvStream, err := tlv.NewStream(
+		tlv.MakePrimitiveRecord(
+			userInitiatedType, &lc.UserInitiated,
+		),
+		tlv.MakePrimitiveRecord(
+			linkFailureErrorType,
+			&linkFailureErrorByte,
+		),
+		tlv.MakeDynamicRecord(
+			htlcActionsType, &lc.HtlcActions,
+			HtlcActionRecordSize(
+				&lc.HtlcActions),
+			HtlcActionEncoder,
+			HtlcActionDecoder,
+		),
+	)
+	var b bytes.Buffer
+	err = tlvStream.Encode(&b)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(w, byteOrder, uint64(b.Len()))
```

Comment:
I think the main reason for varints is in the communication of data (between peers over the wire where every byte counts). So I guess because we are only storing this data and then presenting it to the user via rpc calls we can do it both ways here. No strong opinions here ...